Checker crashes when invoking a function of an undeclared variable

### Problem

The following program crashes the checker:

```swift
let x = y.isInstance(Type<Int>())
```

with

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x14d97ea]

goroutine 15 [running]:
testing.tRunner.func1.1(0x1742f40, 0x1fee2e0)
	/usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:940 +0x2f5
testing.tRunner.func1(0xc0000cc7e0)
	/usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:943 +0x3f9
panic(0x1742f40, 0x1fee2e0)
	/usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/panic.go:969 +0x166
github.com/onflow/cadence/runtime/sema.(*Checker).recordResourceInvalidation(0xc000aa8100, 0x1997ea0, 0xc00028f710, 0x0, 0x0, 0x2, 0x0)
	/Users/bastian/Documents/work/cadence/runtime/sema/checker.go:1411 +0x3a
github.com/onflow/cadence/runtime/sema.(*Checker).checkInvocationExpression(0xc000aa8100, 0xc000278960, 0x0, 0x0)
	/Users/bastian/Documents/work/cadence/runtime/sema/check_invocation_expression.go:158 +0x5e0
github.com/onflow/cadence/runtime/sema.(*Checker).VisitInvocationExpression(0xc000aa8100, 0xc000278960, 0xc000aa8100, 0x2795e10)
	/Users/bastian/Documents/work/cadence/runtime/sema/check_invocation_expression.go:27 +0x3f
github.com/onflow/cadence/runtime/ast.(*InvocationExpression).AcceptExp(...)
	/Users/bastian/Documents/work/cadence/runtime/ast/expression.go:339
github.com/onflow/cadence/runtime/ast.(*InvocationExpression).Accept(0xc000278960, 0x19b8640, 0xc000aa8100, 0xb, 0x2)
	/Users/bastian/Documents/work/cadence/runtime/ast/expression.go:335 +0x62
github.com/onflow/cadence/runtime/sema.(*Checker).visitVariableDeclaration(0xc000aa8100, 0xc0000d58c0, 0x0)
	/Users/bastian/Documents/work/cadence/runtime/sema/check_variable_declaration.go:43 +0xc2
github.com/onflow/cadence/runtime/sema.(*Checker).VisitVariableDeclaration(0xc000aa8100, 0xc0000d58c0, 0x10, 0xc0000bc3e8)
	/Users/bastian/Documents/work/cadence/runtime/sema/check_variable_declaration.go:27 +0x3a
github.com/onflow/cadence/runtime/ast.(*VariableDeclaration).Accept(0xc0000d58c0, 0x19b8640, 0xc000aa8100, 0x1, 0x0)
	/Users/bastian/Documents/work/cadence/runtime/ast/variable_declaration.go:51 +0x3e
github.com/onflow/cadence/runtime/sema.(*Checker).VisitProgram(0xc000aa8100, 0xc0000bc370, 0x0, 0x0)
	/Users/bastian/Documents/work/cadence/runtime/sema/checker.go:453 +0x417
github.com/onflow/cadence/runtime/ast.(*Program).Accept(...)
	/Users/bastian/Documents/work/cadence/runtime/ast/program.go:53
github.com/onflow/cadence/runtime/sema.(*Checker).Check.func1(...)
	/Users/bastian/Documents/work/cadence/runtime/sema/checker.go:326
github.com/onflow/cadence/runtime/sema.(*Checker).Check(0xc000aa8100, 0x1977580, 0x1960f70)
	/Users/bastian/Documents/work/cadence/runtime/sema/checker.go:331 +0x154
github.com/onflow/cadence/runtime/tests/checker.ParseAndCheckWithOptions(0xc0000cc7e0, 0x1852f87, 0x35, 0x0, 0x1977580, 0x1960f70, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/bastian/Documents/work/cadence/runtime/tests/checker/utils.go:95 +0x232
github.com/onflow/cadence/runtime/tests/checker.ParseAndCheck(...)
	/Users/bastian/Documents/work/cadence/runtime/tests/checker/utils.go:36
github.com/onflow/cadence/runtime/tests/checker.TestCheckResourceMoveMemberInvocation.func6(0xc0000cc7e0)
	/Users/bastian/Documents/work/cadence/runtime/tests/checker/resources_test.go:4510 +0x89
testing.tRunner(0xc0000cc7e0, 0x18998c8)
	/usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:991 +0xdc
created by testing.(*T).Run
	/usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:1042 +0x357
```

### Notes

Discovered using the fuzzer.

The problem occurs when invoking a member of an undeclared variable

--- a/runtime/sema/check_invocation_expression.go
+++ b/runtime/sema/check_invocation_expression.go
@@ -134,6 +134,22 @@ func (checker *Checker) checkInvocationExpression(invocationExpression *ast.Invo
 		inCreate,
 	)
 
+	checker.checkMemberInvocationResourceInvalidation(invokedExpression)
+
+	// Update the return info for invocations that do not return (i.e. have a `Never` return type)
+
+	if _, ok = returnType.(*NeverType); ok {
+		functionActivation := checker.functionActivations.Current()
+		functionActivation.ReturnInfo.DefinitelyHalted = true
+	}
+
+	if isOptionalChainingResult {
+		return &OptionalType{Type: returnType}
+	}
+	return returnType
+}
+
+func (checker *Checker) checkMemberInvocationResourceInvalidation(invokedExpression ast.Expression) {
 	// If the invocation is on a resource, i.e., a member expression where the accessed expression
 	// is an identifier which refers to a resource, then the resource is temporarily "moved into"
 	// the function and back out after the invocation.
@@ -147,47 +163,42 @@ func (checker *Checker) checkInvocationExpression(invocationExpression *ast.Invo
 	// the arguments might just use to the temporarily moved resource, e.g. `foo.bar(foo.baz)`
 	// and not invalidate it.
 
-	if invokedMemberExpression, ok := invokedExpression.(*ast.MemberExpression); ok {
-		if invocationIdentifierExpression, ok := invokedMemberExpression.Expression.(*ast.IdentifierExpression); ok {
-
-			// Check that an entry for `IdentifierInInvocationTypes` exists,
-			// because the entry might be missing if the invocation was on a non-existent variable
-
-			if valueType, ok := checker.Elaboration.IdentifierInInvocationTypes[invocationIdentifierExpression]; ok {
-
-				invalidation := checker.recordResourceInvalidation(
-					invocationIdentifierExpression,
-					valueType,
-					ResourceInvalidationKindMoveTemporary,
-				)
-
-				if invalidation != nil {
-
-					checker.resources.RemoveTemporaryInvalidation(
-						invalidation.resource,
-						invalidation.invalidation,
-					)
-
-					checker.checkResourceUseAfterInvalidation(
-						invalidation.resource,
-						invocationIdentifierExpression,
-					)
-				}
-			}
-		}
+	invokedMemberExpression, ok := invokedExpression.(*ast.MemberExpression)
+	if !ok {
+		return
+	}
+	invocationIdentifierExpression, ok := invokedMemberExpression.Expression.(*ast.IdentifierExpression)
+	if !ok {
+		return
 	}
 
-	// Update the return info for invocations that do not return (i.e. have a `Never` return type)
+	// Check that an entry for `IdentifierInInvocationTypes` exists,
+	// because the entry might be missing if the invocation was on a non-existent variable
 
-	if _, ok = returnType.(*NeverType); ok {
-		functionActivation := checker.functionActivations.Current()
-		functionActivation.ReturnInfo.DefinitelyHalted = true
+	valueType, ok := checker.Elaboration.IdentifierInInvocationTypes[invocationIdentifierExpression]
+	if !ok {
+		return
 	}
 
-	if isOptionalChainingResult {
-		return &OptionalType{Type: returnType}
+	invalidation := checker.recordResourceInvalidation(
+		invocationIdentifierExpression,
+		valueType,
+		ResourceInvalidationKindMoveTemporary,
+	)
+
+	if invalidation == nil {
+		return
 	}
-	return returnType
+
+	checker.resources.RemoveTemporaryInvalidation(
+		invalidation.resource,
+		invalidation.invalidation,
+	)
+
+	checker.checkResourceUseAfterInvalidation(
+		invalidation.resource,
+		invocationIdentifierExpression,
+	)
 }
 
 func (checker *Checker) checkConstructorInvocationWithResourceResult(

--- a/runtime/sema/check_invocation_expression.go
+++ b/runtime/sema/check_invocation_expression.go
@@ -150,25 +150,29 @@ func (checker *Checker) checkInvocationExpression(invocationExpression *ast.Invo
 	if invokedMemberExpression, ok := invokedExpression.(*ast.MemberExpression); ok {
 		if invocationIdentifierExpression, ok := invokedMemberExpression.Expression.(*ast.IdentifierExpression); ok {
 
-			valueType := checker.Elaboration.IdentifierInInvocationTypes[invocationIdentifierExpression]
+			// Check that an entry for `IdentifierInInvocationTypes` exists,
+			// because the entry might be missing if the invocation was on a non-existent variable
 
-			invalidation := checker.recordResourceInvalidation(
-				invocationIdentifierExpression,
-				valueType,
-				ResourceInvalidationKindMoveTemporary,
-			)
+			if valueType, ok := checker.Elaboration.IdentifierInInvocationTypes[invocationIdentifierExpression]; ok {
 
-			if invalidation != nil {
-
-				checker.resources.RemoveTemporaryInvalidation(
-					invalidation.resource,
-					invalidation.invalidation,
-				)
-
-				checker.checkResourceUseAfterInvalidation(
-					invalidation.resource,
+				invalidation := checker.recordResourceInvalidation(
 					invocationIdentifierExpression,
+					valueType,
+					ResourceInvalidationKindMoveTemporary,
 				)
+
+				if invalidation != nil {
+
+					checker.resources.RemoveTemporaryInvalidation(
+						invalidation.resource,
+						invalidation.invalidation,
+					)
+
+					checker.checkResourceUseAfterInvalidation(
+						invalidation.resource,
+						invocationIdentifierExpression,
+					)
+				}
 			}
 		}
 	}

--- a/runtime/tests/checker/resources_test.go
+++ b/runtime/tests/checker/resources_test.go
@@ -4502,4 +4502,17 @@ func TestCheckResourceMoveMemberInvocation(t *testing.T) {
 
 		assert.IsType(t, &sema.ResourceLossError{}, errs[0])
 	})
+
+	t.Run("invocation on undeclared variable", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+          let x = y.isInstance(Type<Int>())
+        `)
+
+		errs := ExpectCheckerErrors(t, err, 1)
+
+		assert.IsType(t, &sema.NotDeclaredError{}, errs[0])
+	})
 }

